figma/lib/helpers.ts

```diff
@@ -0,0 +1,143 @@
+        extractors.forEach((fn) => {
+          const res = fn(child) as ReturnType<F>
+          if (!res) return
+          onDataFound(res, parent)
+        })
+        if (!('children' in child) || !child.children?.length) return
+        walk(child.children, child)
+      })
+    }
+  }
+}
+
+export function getFileTopLevelChildren(file: Node<'DOCUMENT'>): Array<Node<'CANVAS'>> {
+  return file.children.filter(
+    (child) => ['CANVAS'].includes(child.type) && 'children' in child && child.children?.length
+  ) as Array<Node<'CANVAS'>>
+}
+
+type OperatorFunction<T, T1> = (arg: T) => T1
+
+export function pipe<T>(): (arg: T) => T
+export function pipe<T, A>(op1: OperatorFunction<T, A>): (arg: T) => A
+export function pipe<T, A, B>(op1: OperatorFunction<T, A>, op2: OperatorFunction<A, B>): (arg: T) => B
+export function pipe<T, A, B, C>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>
+): (arg: T) => C
+export function pipe<T, A, B, C, D>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>
+): (arg: T) => D
+export function pipe<T, A, B, C, D, E>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>,
+  op5: OperatorFunction<D, E>
+): (arg: T) => E
+export function pipe<T, A, B, C, D, E, F>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>,
+  op5: OperatorFunction<D, E>,
+  op6: OperatorFunction<E, F>
+): (arg: T) => F
+export function pipe<T, A, B, C, D, E, F, G>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>,
+  op5: OperatorFunction<D, E>,
+  op6: OperatorFunction<E, F>,
+  op7: OperatorFunction<F, G>
+): (arg: T) => G
+export function pipe<T, A, B, C, D, E, F, G, H>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>,
+  op5: OperatorFunction<D, E>,
+  op6: OperatorFunction<E, F>,
+  op7: OperatorFunction<F, G>,
+  op8: OperatorFunction<G, H>
+): (arg: T) => H
+export function pipe<T, A, B, C, D, E, F, G, H, I>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>,
+  op5: OperatorFunction<D, E>,
+  op6: OperatorFunction<E, F>,
+  op7: OperatorFunction<F, G>,
+  op8: OperatorFunction<G, H>,
+  op9: OperatorFunction<H, I>
+): (arg: T) => I
+export function pipe<T, A, B, C, D, E, F, G, H, I>(
+  op1: OperatorFunction<T, A>,
+  op2: OperatorFunction<A, B>,
+  op3: OperatorFunction<B, C>,
+  op4: OperatorFunction<C, D>,
+  op5: OperatorFunction<D, E>,
+  op6: OperatorFunction<E, F>,
+  op7: OperatorFunction<F, G>,
+  op8: OperatorFunction<G, H>,
+  op9: OperatorFunction<H, I>,
+  ...operations: OperatorFunction<any, any>[]
+): unknown
```

Comment:
Unfortunately no.
One of the most annoying restrictions of typescript is the lack of recursion in generics, so you have to have these overloads. As for the functionality it provides - basically it's all about providing an interface, where you give this `pipe` a list of functions taking an argument and returning a value and it will type-check them all, and infer correct initial parameter and result, so:
```ts
const result = pipe((a: string) => parseInt(a, 10),
        (b: string) => 'whatever))('10') 
// this will fail because output of the first lambda does not match the input of the second
```
I wanted to do generics during this typescript workshop, because I feel like it's the most "alien" thing about typescript